pkg/zng/docs/spec.md

```diff
@@ -1,92 +1,366 @@
+"the next available" integer type code to a type encoding.  Type codes
+begin at the value 23 and increase by one for each typedef. This bindings
+are scoped to the stream in which the typedef occurs.
+
+Type codes for the "scalar types" are predefined as follows:
+
+| Type     | Code |
+|----------|------|
+| bool     |   0  |
+| byte     |   1  |
+| int16    |   2  |
+| uint16   |   3  |
+| int32    |   4  |
+| uint32   |   5  |
+| int64    |   6  |
+| uint64   |   7  |
+| float64  |   8  |
+| string   |   9  |
+| bytes    |  10  |
+| bstring  |  11  |
+| enum     |  12  |
+| ip       |  13  |
+| port     |  14  |
+| net      |  15  |
+| time     |  16  |
+| duration |  17  |
+| any      |  18  |
+
+#### BZNG typedef
+
+A typedef defines a new composite type (i.e., record, array, or set)
+from the composite type code plus the composite type-specific encoding of the
+new type.
+
+The composite type codes are:
+
+| Type     | Code |
+|----------|------|
+| record   |   0  |
+| array    |   1  |
+| set      |   2  |
+
+Note that the composite type codes overlap with the scalar type codes because their
+uses never overlap.  Composite type codes can only appear at the beginning of
+a typedef.  Thus, the space of referenceable type codes in a given stream is
+the union of the scalar type codes and the type codes created with typedefs.
+
+A typedef is encoded as a single byte indicating the composite type code following by
+the type encoding.  This creates a binding between the implied type code
+(i.e., 23 plus the count of all previous typedefs in the stream) and the new
+type definition.
+
+A "uvarint" is an unsigned integer encoded according to the protobuf specification.
+
+#### Record Type Encoding
 
-Any line beginning with `#` that does not conform with the syntax of a
-directive is an error.
-When errors are encountered parsing ZSON, an implementation should return a
-corresponding error and allow ZSON parsing to proceed if desired.
+Record types consist of an ordered set of columns where each column consists of
+a name and a typed value.  Unlike JSON, the ordering of the columns is significant
+and must be preserved through any APIs that consume, process, and emit ZNG records.
 
-## Regular Directives
+A record type is encoded as a count of fields followed by the field definitions,
+where a field definition is a field name followed by a type code.
 
-Regular directives have just three forms, each with a corresponding structure.
+The field names in a record must be unique.
 
-| Directive     | Example structure                   |
-|---------------|-------------------------------------|
-| Descriptor    | `#<int>:<type>`                     |
-| Ordering hint | `#sort [+-]<field>,[+-]<field>,...` |
-| Control       | `#!<payload>`                       |
+The count of fields is encoded as a uvarint.
 
-### Descriptor Directive
+The field name is encoded as a UTF8 string defining a "ZNG identifier"
```

Comment:
We write "UTF-8" above and start saying "UTF8" here, so I guess we should be consistent. It looks like "UTF-8" is the right way.

Reply:
+1